### PR TITLE
[DOC] Add Quantile forest examples

### LINKED ISSUE #87: Add quantile-regression forest examples
Now the sklearn fork has the ability to predict quantiles. It would be great to demonstrate this ability within `scikit-tree`. The most important examples to replicate would be:

1. https://zillow.github.io/quantile-forest/auto_examples/plot_quantile_vs_standard_forest.html#sphx-glr-auto-examples-plot-quantile-vs-standard-forest-py
2. https://zillow.github.io/quantile-forest/auto_examples/plot_quantile_toy_example.html
3. https://zillow.github.io/quantile-forest/auto_examples/plot_quantile_regression_intervals.html

But honestly, we should replicate all of the examples in `quantile-forest` package, since that package is not really maintained anymore. We can then describe how we achieved feature parity w/ quantile-forest and now support other types of quantile-tree methods such as oblique trees and morf trees.

### PR DESCRIPTION
**Pull request by SUKI-O**
Fixes #87 

Changes proposed in this pull request:

- Add examples to show how to use RandomForestRegressor class to genera quantile predictions.

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

**Comment by adam2392** [comment]
The 4 rendered artifacts: https://output.circle-artifacts.com/output/job/5db6e8d1-7cbe-4795-80d5-bbe5ab46dde6/artifacts/0/dev/use.html

@SUKI-O perhaps we can put this into its own section? I.e. `Quantile Predictions`? See the other example sections. This will help group the examples better.

File: examples/plot_quantile_interpolation_with_RF.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
+"""
+========================================================
+Predicting with different quantile interpolation methods
+========================================================
+
+An example comparison of interpolation methods that can be applied during
+prediction when the desired quantile lies between two data points.
+
+"""
+print(__doc__)
```
Comment (adam2392) [inline examples/plot_quantile_interpolation_with_RF.py]: Unnecessary

File: examples/plot_quantile_regression_intervals_with_RF.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
+"""
+==========================================================
+Quantile prediction intervals with Random Forest Regressor
+==========================================================
+
+An example of how to generate quantile prediction intervals with
+Random Forest Regressor class on the California Housing dataset.
+
+"""
+print(__doc__)
```
Comment (adam2392) [inline examples/plot_quantile_regression_intervals_with_RF.py]: Unnecessary

File: examples/plot_quantile_vs_standard_oblique_forest.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+==============================================================
+Quantile regression vs. standard and oblique regression forest
+==============================================================
+
+An example to generate quantile predictions using an oblique random forest
+instance on a synthetic, right-skewed dataset.
+In a right-skewed distribution, the mean is to the right of the median.
+
+"""
+print(__doc__)
```
Comment (adam2392) [inline examples/plot_quantile_vs_standard_oblique_forest.py]: Unnecessary

File: examples/plot_quantile_toy_example_with_RF.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+Quantile prediction with Random Forest Regressor class
+======================================================
+
+An example that demonstrates how to use the Random Forest to generate
+quantile predictions such as conditional median and prediction intervals.
+The example compares the predictions to a ground truth function used
+to generate noisy samples.
+
+"""
+print(__doc__)
```
Comment (adam2392) [inline examples/plot_quantile_toy_example_with_RF.py]: Unnecessary

**Review by adam2392 — CHANGES_REQUESTED** [review]
Great start and very encouraging that the simulations look very similar to those in `quantile-forests`. 

Can you try to break up each example to make them easier to read? 

For example, see the following example: https://output.circle-artifacts.com/output/job/5db6e8d1-7cbe-4795-80d5-bbe5ab46dde6/artifacts/0/dev/auto_examples/outlier_detection/plot_extended_isolation_forest.html#sphx-glr-auto-examples-outlier-detection-plot-extended-isolation-forest-py

In addition, perhaps we can be more explicit about what each example should show. What is the take-away message? E.g. in the right-skewed simulation, is it that QORF is better than standard ORF? What does better mean?

**Review by sampan501 — CHANGES_REQUESTED** [review]
Looks good after removing the print statements that @adam2392 referenced

**Comment by adam2392** [comment]
https://output.circle-artifacts.com/output/job/eb2e193e-9f05-4d58-903b-41b9a7cf478c/artifacts/0/dev/use.html

I don't see the examples rendered now so there's probably an error in your change. You can leverage the development guide to build the docs locally for yourself to peruse.

**Comment by adam2392** [comment]
You are probably missing a section header which is rendered from the README files inside each subsection

File: doc/whats_new/v0.3.rst [hunk 5]
```diff
@@ -25,4 +25,4 @@ the project since version inception, including:
 
 * `Adam Li`_
 * `Sambit Panda`_
-* `Yuxin Bai`_
```
Comment (adam2392) [inline doc/whats_new/v0.3.rst]: Can you un-revert this change?

File: examples/quantile_predictions/README.txt [hunk 6]
```diff
@@ -0,0 +1,6 @@
+.. _quantile_examples:
+
+Quantile Predictions with Random Forest
+---------------------------------------
+
+Examples demonstrating how to generate quantile predictions using RandomForest class.
```
Comment (adam2392) [inline examples/quantile_predictions/README.txt]:
```suggestion
Examples demonstrating how to generate quantile predictions using Random Forest variants.
```

**Review by adam2392 — CHANGES_REQUESTED** [review]
I know the examples are copied "as-is" from quantile-forests, but I think their documentation is not as comprehensive as I would've liked.

The examples could use a bit more explanation. Otherwise, someone reading these have no idea how to interpret the results, or the figures. WDYT @SUKI-O @sampan501 @PSSF23 ?

File: examples/quantile_predictions/plot_quantile_interpolation_with_RF.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    y_errs.append(
+        np.concatenate(
+            (
+                [y_pred[1] - y_pred[0]],
+                [y_pred[2] - y_pred[1]],
+            ),
+            axis=0,
+        )
+    )
+
```
Comment (sampan501) [inline examples/quantile_predictions/plot_quantile_interpolation_with_RF.py:92]: Maybe add a comment here describing what you are plotting, and a comment at the end explaining how to interpret the figure

File: examples/quantile_predictions/plot_quantile_regression_intervals_with_RF.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+# Scale data to dollars.
+y_true *= 1e5
+y_pred *= 1e5
+y_pred_lower *= 1e5
+y_pred_upper *= 1e5
+
+fig, (ax1, ax2) = plt.subplots(nrows=1, ncols=2, figsize=(10, 4))
+
+y_pred_interval = y_pred_upper - y_pred_lower
```
Comment (sampan501) [inline examples/quantile_predictions/plot_quantile_regression_intervals_with_RF.py:134]: Same as above comment with regard to figures

File: examples/quantile_predictions/plot_quantile_vs_standard_oblique_forest.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        node_to_indices[tree][leaf_ids_test[sample][tree]] for tree in range(leaf_ids_test.shape[1])
+    ]
+    # merge the list of indices into one
+    idx = [item for sublist in li for item in sublist]
+    # get the y_train for each corresponding id
+    y_pred_quantile.append(y_train[idx])
+# get the quatile preditions for each predicted sample
+y_pred_quantile = [np.quantile(y_pred_quantile[i], 0.5) for i in range(len(y_pred_quantile))]
+
+colors = ["#c0c0c0", "#a6e5ff", "#e7a4f5"]
```
Comment (sampan501) [inline examples/quantile_predictions/plot_quantile_vs_standard_oblique_forest.py:85]: Same as above comment with regard to figures

File: examples/quantile_predictions/plot_quantile_toy_example_with_RF.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    # merge the list of indices into one
+    idx = [item for sublist in li for item in sublist]
+    # get the y_train for each corresponding id
+    y_pred_quatile.append(y_train[idx])
+# get the quatile preditions for each predicted sample
+y_pred_low = [np.quantile(y_pred_quatile[i], 0.025) for i in range(len(y_pred_quatile))]
+y_pred_med = [np.quantile(y_pred_quatile[i], 0.5) for i in range(len(y_pred_quatile))]
+y_pred_upp = [np.quantile(y_pred_quatile[i], 0.975) for i in range(len(y_pred_quatile))]
+
+plt.plot(X_test, y_test, ".", c="#f2a619", label="Test Observations", ms=5)
```
Comment (sampan501) [inline examples/quantile_predictions/plot_quantile_toy_example_with_RF.py:91]: Same as above comment with regard to figures

**Review by sampan501 — CHANGES_REQUESTED** [review]
I think the only thing missing is a few comments about what exactly you are plotting and how to interpret the plots and we should be good to go

File: examples/quantile_predictions/plot_quantile_vs_standard_oblique_forest.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+regr_orf.fit(X_train, y_train)
+
+y_pred_orf = regr_orf.predict(X_test)
+# %%
+# Generate Quantile Predictions
+# -----------------------------
+# The idea is for each prediction, the training samples that fell into the same leaf nodes
+# are collected then used to generate the quantile statistics for the desired prediction.
+
+# Fit the model to the training samples
```
Comment (adam2392) [inline examples/quantile_predictions/plot_quantile_vs_standard_oblique_forest.py]:
```suggestion
# Get the leaf-nodes the training samples fall into
```

File: examples/quantile_predictions/plot_quantile_interpolation_with_RF.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+# generated using different interpolation methods.
+
+X = np.array([[-1, -1], [-1, -1], [-1, -1], [1, 1], [1, 1]])
+y = np.array([-2, -1, 0, 1, 2])
+
+# %%
+# The interpolation methods
+# -------------------------
+# The following interpolation methods demonstrated here are:
+# To interpolate between the data points, i and j (i<=j),
```
Comment (adam2392) [inline examples/quantile_predictions/plot_quantile_interpolation_with_RF.py]:
```suggestion
# To interpolate between the data points, i and j (``i <= j``),
```

File: examples/quantile_predictions/plot_quantile_interpolation_with_RF.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+# 1. linear: i + (j - i)*fraction, where fraction is the fractional part of
+# the index surrounded by i and j.
+# 2. lower:  i
+# 3. higher: j
+# 4. midpoint: (i + j)/2.
+# 5. nearest: i or j whichever is nearest.
```
Comment (adam2392) [inline examples/quantile_predictions/plot_quantile_interpolation_with_RF.py]:
```suggestion
# linear, lower, higher, midpoint, or nearest. For more details, see `sktree.RandomForestRegressor`.
```

**Review by adam2392 — CHANGES_REQUESTED** [review]
Few minor changes and then LGTM. 

Generally, it's useful to render the docs locally to see if there's any formatting issues.

Xref: https://output.circle-artifacts.com/output/job/b7f1c854-cb46-4c2d-8a79-57965fe88bd9/artifacts/0/dev/use.html#quantile-predictions-with-random-forest

**Review by adam2392 — APPROVED** [review]

**Review by sampan501 — APPROVED** [review]

**Pull request merged**